### PR TITLE
[Model] Snowflake arctic model implementation

### PR DESCRIPTION
**Pull request by sfc-gh-hazhang**
FILL IN THE PR DESCRIPTION HERE

* Implementation of the `snowflake/arctic` model
* in order to support this model, we need the integration of deepspeedfp8 quantization, which is provided in this PR
* We also improve the model loading logic in vllm with a new `load_format = state_dict`, which is able to load weights from presharded weights (tp=8)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by simon-mo** [comment]
@comaniac will review and shepherd

File: vllm/model_executor/layers/quantization/deepspeedfp.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+class DeepSpeedFPParameter(nn.Parameter):
+    """
+    DeepSpeedFP quantized parameter class that implements fp8/fp6
+    quantization deepspeed. Weights are stored in quantized form on
+    GPUs, and can be dequantized on-the-fly when needed by the model.
+    """
+
+    def __new__(cls, orig_shape: torch.Size, params_dtype: torch.dtype,
+                quant_config: DeepSpeedFPConfig):
+        from deepspeed.ops.fp_quantizer import FP_Quantize
```
Comment (comaniac) [inline vllm/model_executor/layers/quantization/deepspeedfp.py:149]: Can we add a comment to note the FP_Quantize version we've tested against to? And it would be better to throw an error with a proper error message when ImportError.

File: vllm/model_executor/model_loader/loader.py [hunk 2]
```diff
@@ -347,6 +347,40 @@ def load_model(self, *, model_config: ModelConfig,
...
+            rank = get_tensor_model_parallel_rank()
+            paths = glob.glob(
+                f"{model_config.model}/model-{rank}-*.safetensors")
+            params = dict(model.named_parameters())
+            for file in paths:
+                for key, val in load_file(file).items():
+                    with torch.no_grad():
+                        params[key].copy_(val)
+                    params.pop(key)
+            assert not params
```
Comment (comaniac) [inline vllm/model_executor/model_loader/loader.py:380]: IIUC, unlike default loader that shards weights on the fly, the current state dict loader only supports pre-sharded checkpoints. Is it correct? If so, it would be better to support general checkpoints as StaticDictLoader can be a common loader for different use cases. Another option is adding checkers and notes:
- When `model-{rank}-*.safetensors` is missing, throw an error saying the state dict is expected to be sharded in advance.
- Add a TODO saying we should support un-sharded state dicts.

File: vllm/worker/model_runner.py [hunk 3]
```diff
@@ -212,6 +212,26 @@ def load_model(self) -> None:
                            "but the KV cache data type is not FP8. "
                            "KV cache scaling factors will not be used.")
 
+    def save_model(self, path: str, max_size: int) -> None:
```
Comment (comaniac) [inline vllm/worker/model_runner.py:215]: Again this name is too general so it seems not proper to save the model in this way. We may call it `save_sharded_state_dicts` for now given its functionality.

File: vllm/worker/worker.py [hunk 4]
```diff
@@ -117,6 +117,9 @@ def init_device(self) -> None:
     def load_model(self):
         self.model_runner.load_model()
 
+    def save_model(self, path: str, max_size: int) -> None:
+        self.model_runner.save_model(path, max_size=max_size)
```
Comment (comaniac) [inline vllm/worker/worker.py:121]: Ditto on naming.

**Comment by comaniac** [comment]
cc @pcmoritz please also take a look.

File: vllm/model_executor/layers/fused_moe/fused_moe.py [hunk 5]
```diff
@@ -477,3 +460,63 @@ def fused_moe(
...
+                         w2,
+                         topk_weights,
+                         topk_ids,
+                         inplace=inplace,
+                         override_config=override_config,
+                         use_fp8=use_fp8,
+                         w1_scale=w1_scale,
+                         w2_scale=w2_scale,
+                         a1_scale=a1_scale,
+                         a2_scale=a2_scale)
```
[the remark this answers is not in the document]
Comment (pcmoritz) [reply vllm/model_executor/layers/fused_moe/fused_moe.py:522]: I'm ok with this interface change, cc @WoosukKwon who has been working on this code recently too :)

File: vllm/model_executor/model_loader/loader.py [hunk 6]
```diff
@@ -347,6 +347,40 @@ def load_model(self, *, model_config: ModelConfig,
...
+            with torch.device(device_config.device):
+                model = _initialize_model(model_config, self.load_config,
+                                          lora_config, vision_language_config)
+            rank = get_tensor_model_parallel_rank()
+            paths = glob.glob(
+                f"{model_config.model}/model-{rank}-*.safetensors")
+            params = dict(model.named_parameters())
+            for file in paths:
+                for key, val in load_file(file).items():
+                    with torch.no_grad():
```
Comment (Yard1) [inline vllm/model_executor/model_loader/loader.py:377]: move context manager outside the loops

File: vllm/worker/model_runner.py (hunk 3 above)
Comment (Yard1) [reply vllm/worker/model_runner.py:215]: would probably make even more sense to make this a part of the `StateDictLoader`

File: vllm/model_executor/model_loader/loader.py [hunk 7]
```diff
@@ -347,6 +347,40 @@ def load_model(self, *, model_config: ModelConfig,
                                              vision_language_config)
 
 
+class StateDictLoader(BaseModelLoader):
```
Comment (Yard1) [inline vllm/model_executor/model_loader/loader.py:350]: add docstring

File: vllm/model_executor/layers/quantization/deepspeedfp.py (hunk 1 above)
Comment (jeffra) [reply vllm/model_executor/layers/quantization/deepspeedfp.py:149]: `FP_Quantize` requires `deepspeed>=0.14.2`. Is there a current way to indicate this type of dependency in vllm?

Comment (comaniac) [reply vllm/model_executor/layers/quantization/deepspeedfp.py:149]: This is now a soft/lazy dependency so we don't put it in `requirements.txt`. What I meant was just add a checker here like

```python
try:
  from deepspeed.ops.fp_quantizer import FP_Quantize
  if FP_Quantize.__version__ < 0.14.2:
    raise ImportError("wrong version")
except ImportError as err:
    raise ImportError(f"nicer message from {err}") from None
```

Comment (jeffra) [reply vllm/model_executor/layers/quantization/deepspeedfp.py:149]: Ah yeah, no worries. The reason I asked is `transformers` has a few nice utils that are used throughout their codebase for optional/lazy dependencies like this (examples below). If there's not then would be nice to use something like `packaging` for the version comparison if possible. 

* https://github.com/huggingface/transformers/blob/5b7a225f2502840f4066b1b39ef53411ea4d6e90/src/transformers/utils/versions.py#L49
* https://github.com/huggingface/transformers/blob/5b7a225f2502840f4066b1b39ef53411ea4d6e90/src/transformers/integrations/deepspeed.py#L34-L47

File: vllm/model_executor/layers/quantization/deepspeedfp.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+        extra_weight_attrs["weight_loader"] = quant_weight_loader
+        set_weight_attrs(weight, extra_weight_attrs)
+
+    def apply(self,
+              layer: torch.nn.Module,
+              x: torch.Tensor,
+              bias: Optional[torch.Tensor] = None) -> torch.Tensor:
+        weight = layer.weight
+        y = weight.ds_dequantize()
+        return F.linear(x, y, bias)
```
[the remark this answers is not in the document]
Comment (aurickq) [reply vllm/model_executor/layers/quantization/deepspeedfp.py:137]: yep that's correct

File: vllm/model_executor/layers/quantization/deepspeedfp.py (hunk 1 above)
Comment (comaniac) [reply vllm/model_executor/layers/quantization/deepspeedfp.py:149]: I don't have much content on this. Maybe @simon-mo can comment more.

Comment (sfc-gh-hazhang) [reply vllm/model_executor/layers/quantization/deepspeedfp.py]: fixed.

File: vllm/model_executor/layers/quantization/deepspeedfp.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+import torch.nn.functional as F
+
+from vllm.model_executor.layers.linear import LinearBase, LinearMethodBase
+from vllm.model_executor.layers.quantization.base_config import (
+    QuantizationConfig)
+from vllm.model_executor.utils import set_weight_attrs
+
+
+class DeepSpeedFPConfig(QuantizationConfig):
+    """Config for DeepSpeed FP quantizer. It supports fp6 and fp8."""
```
[the remark this answers is not in the document]
Comment (sfc-gh-hazhang) [reply vllm/model_executor/layers/quantization/deepspeedfp.py]: done!

**Comment by sfc-gh-hazhang** [comment]
@sfc-gh-aqiao will address the rest related with save/load etc.

File: vllm/model_executor/model_loader/loader.py (hunk 2 above)
Comment (aurickq) [reply vllm/model_executor/model_loader/loader.py]: Done - added checks and todo

File: vllm/worker/model_runner.py (hunk 3 above)
Comment (aurickq) [reply vllm/worker/model_runner.py]: renamed to `save_sharded_state` and moved internal logic to `ShardedStateLoader`

File: vllm/worker/worker.py (hunk 4 above)
Comment (aurickq) [reply vllm/worker/worker.py]: renamed to `save_sharded_state`

File: vllm/model_executor/model_loader/loader.py (hunk 6 above)
Comment (aurickq) [reply vllm/model_executor/model_loader/loader.py]: Done

File: vllm/model_executor/model_loader/loader.py [hunk 10]
```diff
@@ -347,6 +347,40 @@ def load_model(self, *, model_config: ModelConfig,
...
+
+        from vllm.distributed import get_tensor_model_parallel_rank
+        with set_default_torch_dtype(model_config.dtype):
+            with torch.device(device_config.device):
+                model = _initialize_model(model_config, self.load_config,
+                                          lora_config, vision_language_config)
+            rank = get_tensor_model_parallel_rank()
+            paths = glob.glob(
+                f"{model_config.model}/model-{rank}-*.safetensors")
+            params = dict(model.named_parameters())
```
[the remark this answers is not in the document]
Comment (aurickq) [reply vllm/model_executor/model_loader/loader.py]: changed to state_dict

File: vllm/model_executor/model_loader/loader.py (hunk 7 above)
Comment (aurickq) [reply vllm/model_executor/model_loader/loader.py]: Done

**Comment by aurickq** [comment]
Addressed comments and tested. @comaniac ready for another round of feedback and/or approval

File: vllm/model_executor/layers/quantization/deepspeedfp.py [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+    def __init__(
+        self,
+        weight_bits: int = 8,
+        group_size: int = 512,
+    ) -> None:
+        self.weight_bits = weight_bits
+        self.group_size = group_size
+        self.valid_types = [torch.bfloat16, torch.float16]
+
+        if self.weight_bits not in [6, 8]:
```
Comment (comaniac) [inline vllm/model_executor/layers/quantization/deepspeedfp.py:30]: nit: Better to use an immutable tuple instead of a list.
```suggestion
        if self.weight_bits not in (6, 8):
```

**Review by comaniac — APPROVED** [review]
LGTM. Just a nit.
I'll leave to @Yard1 to approve and merge.

**Review by zhuohan123 — CHANGES_REQUESTED** [review]
This PR looks good on the model and quantization part, but can we move the `sharded_state` related logic to a separate PR? I think we need more careful design on that part.

**Comment by aurickq** [comment]
Created separate PR for sharded state loader at https://github.com/vllm-project/vllm/pull/4690

**Comment by sfc-gh-hazhang** [comment]
@simon-mo @WoosukKwon @Yard1 : ready, let's merge? Meanwhile let's move the discussion to #4690 as it is also critical for loading large models (currently, for arctic with 450B, it takes 15 mins to talk if w/o that PR)

File: benchmarks/README.md [hunk 12]
```diff
@@ -5,4 +5,4 @@
 You can download the dataset by running:
 ```bash
 wget https://huggingface.co/datasets/anon8231489123/ShareGPT_Vicuna_unfiltered/resolve/main/ShareGPT_V3_unfiltered_cleaned_split.json
-```
+```
```
Comment (mgoin) [inline benchmarks/README.md:8]: Remove cruft formatting touch

File: vllm/model_executor/models/arctic.py [hunk 13]
```diff
@@ -0,0 +1,523 @@
...
+                    quant_config=quant_config,
+                )
+                torch.cuda.empty_cache()
+                self.w2s = DeepSpeedFPParameter(
+                    torch.Size((self.num_experts, self.hidden_size,
+                                self.intermediate_size)),
+                    params_dtype=params_dtype,
+                    quant_config=quant_config,
+                )
+                torch.cuda.empty_cache()
```
Comment (mgoin) [inline vllm/model_executor/models/arctic.py:129]: Why is this needed?

File: vllm/model_executor/layers/quantization/deepspeedfp.py [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+    Args: 
+        weight_bits: the target quantization bits, 6 or 8.
+        group_size: group size for quantizaiton, default to 128.
+    """
+
+    def __init__(
+        self,
+        weight_bits: int = 8,
+        group_size: int = 512,
+    ) -> None:
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/deepspeedfp.py:25]: Is the only way to control the weight_bits parameter by setting the value in a quant_config.json?

Comment (sfc-gh-hazhang) [reply vllm/model_executor/layers/quantization/deepspeedfp.py:25]: yes, vllm does not seem to have a way to set quant parameters in arguments

File: vllm/model_executor/models/arctic.py (hunk 13 above)
Comment (sfc-gh-hazhang) [reply vllm/model_executor/models/arctic.py:129]: to avoid pytorch's fragmentation. But now maybe this is not needed -- @aurickq  confirm?

File: benchmarks/README.md (hunk 12 above)
Comment (sfc-gh-hazhang) [reply benchmarks/README.md]: fixed

File: vllm/model_executor/models/arctic.py (hunk 13 above)
Comment (aurickq) [reply vllm/model_executor/models/arctic.py:129]: Yep should be OK to remove now, just tested it

**Review by zhuohan123 — APPROVED** [review]
LGTM! Thanks for your contribution!
